Build and statically link sqlite3 via the amalgamation

photex opened the pull request:
Howdy!

I've been working with the sqlite crate and realized that sqlite wasn't getting linked unless it had been installed which wasn't ideal for my needs. 

This PR adds the latest amalgamation and builds a static archive to link with instead. Has the benefit of working on Windows as well (tested with the MSVC toolchain).

I bumped the minor version of the crate because it seems like a big enough difference.

IvanUkhov commented:
Thanks! Looks solid! I have two comments/questions, though.

First, the [guidelines](http://doc.crates.io/build-script.html#-sys-packages) say that it’s a good idea to “probe the current system for [the library] before resorting to building from source.” If SQLite is already installed, why should one compile it again? Personally, I would first call `pkg-config` and, if it failed, would compile. What do you think?

Second, regarding dynamic vs. static linkage, one usually resides to environment variables for that. Have a look at the naming convention in [pkg-config](http://alexcrichton.com/pkg-config-rs/pkg_config/index.html). What are thoughts about this?

photex commented:
I'm cool with that plan. :) 

Perhaps I could also add a feature to always build the amalgamation as well? For me this is what I'll be needing 100% of the time, but for others I can totally understand wanting to match the shared lib that's likely being used by other parts of their system.

Should the build script fallback to using the amalgamation if pkg_config can't find a system library? 

Should I even bother with trying to use the amalgamation to build a shared lib?

IvanUkhov commented:
Yes, it would be nice to have a feature forcing the build script to compile. I’m not sure what the name of that feature should be, but we should try to stick to the naming conventions if they exist. If it was the other way around, it would be called `system`. Perhaps `binded` or something?

Yes, the build script should try `pkg-config` first and only then `gcc`.

No, I don’t think dynamic linking is worth doing if `pkg-config` fails. Let’s link statically in that case.

IvanUkhov commented:
You know, maybe instead of having `src/sqlite.c`, one should just extract the content of `sqlite-amalgamation-3130000.zip` _as is_ into a separate directory called `source`. Those other three files won’t be used, but it’ll be clear what the origin of the source code is and how to update it.

photex commented:
That's a good idea.
Would it make sense to just try and download the archive from SQLite.org directly as an alternative to storing the files in git?

IvanUkhov commented:
I personally wouldn’t do this. Whenever a library doesn’t keep its code under Git, I usually create a Git mirror and add it as a submodule. [Here](https://github.com/copies) is what I’ve collected so far. When packaging for publishing, the source code will end up in the crate itself. So no internet connection will be needed for the compilation, but, of course, the crate will have to get to the host machine somehow first.

IvanUkhov commented:
I can create a mirror under that organization if you want. [Here](https://github.com/stainless-steel/hdf5-sys) is how I typically organize this.

IvanUkhov commented:
[Here](https://github.com/stainless-steel/hdf5-sys) is an example of how I typically organize this.

IvanUkhov commented:
So [here](https://github.com/copies/sqlite) is a mirror. You can add it as a submodule, tag `v3.13.0`.

IvanUkhov commented:
I was wrong about _binded_. The guys on #cargo suggest using _bound_, _contained_, _included_, or _packaged_.  Let’s use _bound_.

Commit:
adding sqlite3 submodule for upstream sources

Commit:
Ability to build sqlite3 and statically link via the amalgamation

Default behavior is to attempt to use the system library, if that isn't
found then sqlite3 will be built as a static lib using the amalgamation.

For applications that always want to use the amalgamation can use the
'bound' feature.

photex commented on build.rs:
```diff
@@ -1,9 +1,16 @@
 
+#[cfg(not(feature = "bound"))]
 fn main() {
+    const NAME: &'static str = "sqlite3";
```
I moved this line here to prevent a 'unused' warning to be emitted when building with the 'bound' feature.

IvanUkhov replied:
Yeah, makes sense.

IvanUkhov commented:
Thanks a lot for the help!

photex commented:
No problem! I'm happy I could contribute something.
